### DISCUSSION TITLE
Non-nullable graphql fields are being marked as required in the generated typescript code

### DISCUSSION THREAD

**Discussion by vidishanevatia** [dotansimha/graphql-code-generator#10144] [Q&A] [answered]
Problem:
In my project, non-nullable fields in the GraphQL schema (e.g., field: String!) are being marked as required in the generated TypeScript types. This behavior enforces the inclusion of all non-nullable fields in the queries and mutations, even if they are not relevant or needed for specific use cases.

Expected Behavior:
Even though fields are non-nullable in the schema, we should be able to exclude certain fields from our queries and mutations if they are not required for that specific operation. The generated TypeScript types should not enforce all non-nullable fields to be required in the code. This would allow for more flexibility in designing queries and mutations, focusing only on the necessary fields.

Is there a way I can achieve this behavior? @dotansimha

**Comment by charpeni** [comment, marked as answer]
The expected behavior you described matches the behavior I would expect, but it also matches my experience so far.

"Full" types will be generated from the GraphQL schema, and they will include all non-nullable fields as required, but those types shouldn't be used as-is as they are 1:1 with their GraphQL schema type.

Instead, GraphQL operations can be composed of colocated fragments (or not) that expose the data requirements of components. Non-nullable fields shouldn't be required as part of the operation, but they will be non-nullable if queried. Full types aren't expected to be imported. Instead, we should rely on type utilities like `FragmentType`, `ResultOf`, and `VariablesOf` to infer the result or variables types (see [Getting a Fragment's type](https://the-guild.dev/graphql/codegen/plugins/presets/preset-client#getting-a-fragments-type)).

We can even go further by conditionally queried fields with [`@skip`](https://spec.graphql.org/October2021/#sec--skip) and [`@include`](https://spec.graphql.org/October2021/#sec--include) directives. Both are supported by the GraphQL Code Generator, which will turn a non-nullable field into an optional field, as it could be included or not.

I don't see why all non-nullable fields would always be required to be queried as part of all GraphQL operations, but it's hard to help further without knowing your current setup. Do you use `client-preset`? Do you use Fragment Masking? What's your GraphQL client?

**Reply by vidishanevatia** [reply]
No, we are not usimg `client-preset` or Fragment masking. And, thanks, this was a very helpful response!
